### PR TITLE
Update Stream Trigger Scaling logic

### PR DESCRIPTION
**Pull request by mapalan**
Currently, the scaling logic for the RedisStreamTrigger uses only the stream length to determine the target number of function instances. If all entries in the stream have been acked but not deleted, this logic will return a large target instance count even though there are no entries left to process.

Redis 7 contains a `lag` field in the [`XINFO GROUPS`](https://redis.io/commands/xinfo-groups/) command that allows a client to see the number of unacked entries in a stream for a consumer group. This is used instead of the stream length.

Redis 6 does not contain that field like Redis 7. As a workaround, this PR adds a method for the scaler to estimate the number of unacked entries. This relies on the default format of the stream entry id being [`<millisecondsTime>-<sequenceNumber>`](https://redis.io/docs/data-types/streams/#entry-ids). [`XINFO GROUPS`](https://redis.io/commands/xinfo-groups/) contains information about the last entry ID that the group received, and we can get the IDs of the first and last entry in the stream from [`XINFO STREAM`](https://redis.io/commands/xinfo-stream/). We approximate the number of entries remaining using time:
```
(last-entry-id - last-delievered-id) / (last-entry-id - first-entry-id) * streamLength
```

RedisStreamScaleMonitor does not get access to the function's LogName, only the ShortName. The function's consumer group is currently based on the LogName, meaning that the ScaleMonitor would not be able to accurately call `XINFO GROUPS` to get the last delievered id. In order to make the separately initialized RedisStreamScaleMonitor and RedisStreamListener use the same consumer group, this PR changes the consumer group of the function to the function's short name, as that is available to both.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs [hunk 1]
```diff
@@ -11,21 +12,71 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
...
+             * Estimate number of remaining entries in the stream using the unixtime field from the entry IDs:
+             * - default ID structure ("unixtime-counter")
+             * - entries are added at a constant rate over time
+            */
+            StreamInfo info = await multiplexer.GetDatabase().StreamInfoAsync(key);
+            string firstId = (string)info.FirstEntry.Id;
+            string lastId = (string)info.LastEntry.Id;
+            string lastDeliveredId = group.LastDeliveredId ?? firstId;
+
+            long firstTimestamp = long.Parse(firstId.Split('-')[0]);
```
Comment (yugangw-msft) [inline src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:66]: If the entity id is not the default format, and we get it wrong, what will be the worst consequence?

Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs]: I've updated the logic to ensure that as long as there are any unprocessed entries in the streams, the value returned will be between 1 and the stream length. If all entries are processed, then it will return 0.
Even if the ID is custom, the ID format `<number>-<counter>` is the same, and the IDs must strictly be increasing as noted [here](https://redis.io/commands/xadd). Even with custom IDs, the estimation logic will give somewhere within the range.

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs [hunk 2]
```diff
@@ -11,21 +12,91 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
...
+            long lastDeliveredTimestamp = long.Parse(lastDeliveredIdSplit[0]);
+
+            if (lastTimestamp == lastDeliveredTimestamp)
+            {
+                // If timestamp is the same, return counter difference
+                long lastCounter = long.Parse(lastIdSplit[1]);
+                long lastDelieveredCoutner = long.Parse(lastDeliveredIdSplit[1]);
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = Math.Min(streamLength, Math.Max(1, lastCounter - lastDelieveredCoutner)),
```
Comment (yugangw-msft) [inline src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: Our of curious, what edge conditions we want to handle here

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs [hunk 3]
```diff
@@ -11,21 +12,91 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
...
+            string[] lastIdSplit = lastId.Split('-');
+            string[] lastDeliveredIdSplit = lastDeliveredId.Split('-');
+            long firstTimestamp = long.Parse(firstIdSplit[0]);
+            long lastTimestamp = long.Parse(lastIdSplit[0]);
+            long lastDeliveredTimestamp = long.Parse(lastDeliveredIdSplit[0]);
+
+            if (lastTimestamp == lastDeliveredTimestamp)
+            {
+                // If timestamp is the same, return counter difference
+                long lastCounter = long.Parse(lastIdSplit[1]);
```
Comment (yugangw-msft) [inline src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:81]: We shall use ulong https://github.com/redis/redis/blob/068051e378ceaea85a7fb011f1271587a17bd0e7/src/stream.h#L11-L14

```c
typedef struct streamID {
    uint64_t ms;        /* Unix time in milliseconds. */
    uint64_t seq;       /* Sequence number. */
} streamID;
```

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 2 above)
Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: This is specifically for the case if the customer was using custom IDs where they kept the `time` part of the ID as constant and only increased the `counter`. In this case, the estimation of remaining entries would be the difference between the counters of the last entry id and last entry delivered id. The counter can be any strictly increasing sequence, so I limit the `Remaining` to between 1 and the streamLength

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs [hunk 4]
```diff
@@ -11,21 +12,91 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
...
+            ulong lastDeliveredTimestamp = ulong.Parse(lastDeliveredIdSplit[0]);
+
+            if (lastTimestamp == lastDeliveredTimestamp)
+            {
+                // If timestamp is the same, return counter difference
+                ulong lastCounter = ulong.Parse(lastIdSplit[1]);
+                ulong lastDelieveredCoutner = ulong.Parse(lastDeliveredIdSplit[1]);
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = Math.Min(streamLength, (long)Math.Max(1, lastCounter - lastDelieveredCoutner)),
```
Comment (yugangw-msft) [inline src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: There is precision loss from ulong to long, though users might unlikely run into it. Given the count of element in stream is `ulong` as defined in Redis, we shall definitely change our library across. I am fine we do it in a separate PR. It is your call

```c
typedef struct stream {
    rax *rax;               /* The radix tree holding the stream. */
    uint64_t length;        /* Current number of elements inside this stream. */
    streamID last_id;       /* Zero if there are yet no items. */
    streamID first_id;      /* The first non-tombstone entry, zero if empty. */
    streamID max_deleted_entry_id;  /* The maximal ID that was deleted. */
    uint64_t entries_added; /* All time count of elements added. */
    rax *cgroups;           /* Consumer groups dictionary: name -> streamCG */
} stream;
```

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 2 above)
Comment (yugangw-msft) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs]: To learn, inaccurate count isn't impacting the correctness of result, even if in case it happens; rather just impact how many worker will be created?

**Review by yugangw-msft — APPROVED** [review]
I am approving. Please create tracking items for GA if addressing it this PR is not efficient

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 2 above)
Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs]: Yes, the value only impacts the number of requested workers and does not change any logic that processes messages from the stream.

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 3 above)
Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs]: Updated to ulong

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 4 above)
Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: I will have to update this for the list length as well [since that is `ulong` as well](https://github.com/redis/redis/blob/38f02349462d5aefa9b25386a130ebd67db1a4de/src/adlist.h#L47), so I will combine both `long` to `ulong` conversions in a separate PR

**Review by yugangw-msft — APPROVED** [review]

File: src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs (hunk 4 above)
Comment (mapalan) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: StackExchange.Redis uses `long` for both [ListLength](https://github.com/StackExchange/StackExchange.Redis/blob/b2694b385c876793a827bd4d2556439983061877/src/StackExchange.Redis/Interfaces/IDatabase.cs#L1027) and [StreamLength](https://github.com/StackExchange/StackExchange.Redis/blob/b2694b385c876793a827bd4d2556439983061877/src/StackExchange.Redis/Interfaces/IDatabase.cs#L2423C33-L2423C33), so I think it would be best to stay consistent with the SDK and leave it as is

Comment (yugangw-msft) [reply src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs:85]: Thanks. @marcgravell @philon-msft , could you please provide any context here?
